src/transport/raweth/rx.c

```diff
@@ -0,0 +1,365 @@
+//
+// Copyright (c) 2022 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+// #include "zenoh-pico/transport/link/rx.h"
+
+#include <stddef.h>
+
+#include "zenoh-pico/config.h"
+#include "zenoh-pico/protocol/codec/network.h"
+#include "zenoh-pico/protocol/codec/transport.h"
+#include "zenoh-pico/protocol/core.h"
+#include "zenoh-pico/protocol/iobuf.h"
+#include "zenoh-pico/session/utils.h"
+#include "zenoh-pico/transport/utils.h"
+#include "zenoh-pico/utils/logging.h"
+
+#if Z_FEATURE_RAWETH_TRANSPORT == 1
+
+static size_t _z_raweth_link_recv_zbuf(const _z_link_t *link, _z_zbuf_t *zbf, _z_bytes_t *addr) {
+    uint8_t *buff = _z_zbuf_get_wptr(zbf);
+    size_t rb = _z_receive_raweth(&link->_socket._raweth._sock, buff, _z_zbuf_space_left(zbf), addr);
+    // Check validity
+    if ((rb == SIZE_MAX) || (rb < sizeof(_zp_eth_header_t))) {
+        return rb;
+    }
+    // Check if header has vlan
+    _Bool has_vlan = false;
+    _zp_eth_header_t *header = (_zp_eth_header_t *)buff;
+    if (header->length > _ZP_ETH_TYPE_CUTOFF) {
+        has_vlan = true;
+    }
+    // Check validity
+    if (has_vlan && (rb < sizeof(_zp_eth_vlan_header_t))) {
+        return rb;
+    }
+    // Update buffer but skip eth header
+    _z_zbuf_set_wpos(zbf, _z_zbuf_get_wpos(zbf) + rb);
+    if (has_vlan) {
+        _z_zbuf_set_rpos(zbf, _z_zbuf_get_wpos(zbf) - (rb - sizeof(_zp_eth_vlan_header_t)));
+    } else {
+        _z_zbuf_set_rpos(zbf, _z_zbuf_get_wpos(zbf) - (rb - sizeof(_zp_eth_header_t)));
+    }
+    return rb;
+}
+
+_z_transport_peer_entry_t *_z_find_peer_entry(_z_transport_peer_entry_list_t *l, _z_bytes_t *remote_addr) {
+    _z_transport_peer_entry_t *ret = NULL;
+
+    _z_transport_peer_entry_list_t *xs = l;
+    for (; xs != NULL; xs = _z_transport_peer_entry_list_tail(xs)) {
+        _z_transport_peer_entry_t *val = _z_transport_peer_entry_list_head(xs);
+        if (val->_remote_addr.len != remote_addr->len) {
+            continue;
+        }
+
+        if (memcmp(val->_remote_addr.start, remote_addr->start, remote_addr->len) == 0) {
+            ret = val;
+        }
+    }
+    return ret;
+}
+
+/*------------------ Reception helper ------------------*/
+int8_t _z_raweth_recv_t_msg_na(_z_transport_multicast_t *ztm, _z_transport_message_t *t_msg, _z_bytes_t *addr) {
```

Comment:
Regarding code replication, I believe we could move this code in some common structure between multicast and raweth? I believe the really required function is `_z_raweth_link_recv_zbuf` that should perform the right filtering and checking.